File: 2251-number-of-flowers-in-full-bloom/2251-number-of-flowers-in-full-bloom.py

class Solution:
    def fullBloomFlowers(self, flowers: List[List[int]], people: List[int]) -> List[int]:
        
        flower_sweep = {}
        
        for flower in flowers:
            
            if flower[0] not in flower_sweep:
                flower_sweep[flower[0]] = 1
            else:
                flower_sweep[flower[0]] += 1
                
            if flower[1] + 1 not in flower_sweep:
                flower_sweep[flower[1] + 1] = -1
            else:
                flower_sweep[flower[1] + 1] -= 1
        
        sorted_keys = sorted(flower_sweep.keys())
        # print(flower_sweep)
        # print(sorted_keys)
        people_list = [(idx,p) for idx,p in enumerate(people)]
        people_list.sort(key=lambda x: x[1])
        # print(people_list)
        n = len(sorted_keys)
        m = len(people_list)
        f_count = 0
        i,j = 0,0
        ans = [0]*m
        while i < n and j < m:
            if sorted_keys[i] <= people_list[j][1]:
                f_count += flower_sweep[sorted_keys[i]]
                i += 1
            else:
                ans[people_list[j][0]] = f_count
                j += 1
        
        while j < m:
            ans[people_list[j][0]] = f_count
            j += 1
        
        
        return ans
        

File: 2251-number-of-flowers-in-full-bloom/2251-number-of-flowers-in-full-bloom.cpp

int reverse_lb(vector<pair<int,int>>&v,int target)
{
    int l = 0;
    int r = v.size() - 1;
    int ans = -1;
    while(l <= r)
    {
        int mid = (l + r)/2;
        
        if(v[mid].first <= target)
        {
            ans = mid;
            l = mid + 1;
        }
        else
        {
            r = mid - 1;
        }
    }
    
    return ans;
}
class Solution {
public:
    vector<int> fullBloomFlowers(vector<vector<int>>& flowers, vector<int>& people) {
        
        map<int,int>flower_sweep;
        for (auto &flower:flowers)
        {
            flower_sweep[flower[0]]++;
            flower_sweep[flower[1] + 1]--;
        }
        
        int f_count = 0;
        
        vector<pair<int,int>>v;
        
        for (auto &m : flower_sweep)
        {
            f_count += m.second;
           v.push_back({m.first,f_count});
        }
        
        
        vector<int>res;
        for(auto &p : people)
        {
            int pos = reverse_lb(v,p);
            if(pos == -1)
            {
                res.push_back(0);
            }
            else
            {
                res.push_back(v[pos].second);
            }
        }
        
        return res;
    }
};